### PR TITLE
Lo L1A Histogram packet_to_dataset() Refactor and CDF Creation

### PR DESCRIPTION
**Pull request by sdhoyt**
# Change Summary

## Overview
This PR refactors the the Lo L1A histogram code to use the `packet_to_dataset()` function and adds code to create the L1A CDF. There ended up being more changes that I had hoped, so I can try to break this up if needed. What I was doing on this branch evolved a bit over time, so this isn't a compartmentalized as a should be.

## New Dependencies
None

## New Files
- [lo_science.py](https://github.com/IMAP-Science-Operations-Center/imap_processing/compare/dev...sdhoyt:lo-hist-redo?expand=1#diff-8d72285ebf89cb7721cda2846a06a46361dd1b942265d2f7b9581901d89b9909)
   - new L1A science processing file. Currently only contains histogram processing

## Deleted Files
- [science_counts.py](https://github.com/IMAP-Science-Operations-Center/imap_processing/compare/dev...sdhoyt:lo-hist-redo?expand=1#diff-0d7b8001d2375d4dd41e6009bb665b4356542146b3abdc2672ca69d765d733cf)
   - science counts data class not needed with `packets_to_dataset()`
- [lo_data_container.py](https://github.com/IMAP-Science-Operations-Center/imap_processing/compare/dev...sdhoyt:lo-hist-redo?expand=1#diff-08d7be0d488670214ca66eef1fb99a83fad1c4fb314fe2ecb9320f9208b1fcae)
   - data container for data classes no longer needed with `packets_to_dataset()`
- [lo_l1a_write_cdfs.py](https://github.com/IMAP-Science-Operations-Center/imap_processing/compare/dev...sdhoyt:lo-hist-redo?expand=1#diff-4fba08898e5157983ff74cdeba39fc99bee1319043dba109656a3db0a086c012)
   - Planning to set CDF attributes in lo_l1a.py main processing file. No longer need this file.

## Updated Files
- [imap_lo_l1a_variable_attrs.yaml](https://github.com/IMAP-Science-Operations-Center/imap_processing/compare/dev...sdhoyt:lo-hist-redo?expand=1#diff-dece8fbab8dad36da87159084cca38e304c018bf7792a47ec64edd98f07872bc)
   - renamed tof histogram attributes so they don't conflict with DE tof attributes
- [cli.py](https://github.com/IMAP-Science-Operations-Center/imap_processing/compare/dev...sdhoyt:lo-hist-redo?expand=1#diff-82b37771721d671205543c076fa7df7c403ebd277605353b656a26981957e4e2)
   - removed list around processing function return
- [lo_apid.py](https://github.com/IMAP-Science-Operations-Center/imap_processing/compare/dev...sdhoyt:lo-hist-redo?expand=1#diff-be07dccb80388f01281d9cf77e922567b2269d524b760b4b333d424a398a2de0)
   - removed needed APIDs
- [lo_base.py](https://github.com/IMAP-Science-Operations-Center/imap_processing/compare/dev...sdhoyt:lo-hist-redo?expand=1#diff-b913f0869ef301d7477973265f1b737a39ee137ed26fc21577cd116d1196f8aa)
   - updated space packet parser import. This file will eventually be removed once everything is converted to use packets_to_dataset.
- [lo_l1a.py](https://github.com/IMAP-Science-Operations-Center/imap_processing/compare/dev...sdhoyt:lo-hist-redo?expand=1#diff-8b95d157a01299bce0daf049b6edb980f800e7f7cc9b0080dd2cbb3479d1fc7a)
   - Refactored to use packets_to_dataset, removed temporary code from SIT-3, and added cdf attribute creation for L1A hist
- [lo_xtce.xml](https://github.com/IMAP-Science-Operations-Center/imap_processing/compare/dev...sdhoyt:lo-hist-redo?expand=1#diff-e33cf3612a64ca8c41bfe0a5ff507772681ab632fed85b4b22274a047c375f66)
   - fixed name of sci count field.
## Testing

- Updated [test_lo_l1a.py](https://github.com/IMAP-Science-Operations-Center/imap_processing/compare/dev...sdhoyt:lo-hist-redo?expand=1#diff-f93c1c44a3ee1f7d319fd7f92a79768a3b51c3b051a0120e31ce3f8525dc27cf) and [test_science_counts.py](https://github.com/IMAP-Science-Operations-Center/imap_processing/compare/dev...sdhoyt:lo-hist-redo?expand=1#diff-297d274055a990f59ebd214283b7fc7dd911baf118e48ffa06c88fb95475999c). 
- Tested with cli.py

### CODE REVIEW

File: imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml
Comment (greglucas) [inline imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml]: Suggest this being a `uint8` datatype and FILLVAL being 255 then.

File: imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml
Comment (greglucas) [inline imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml:45]: I think this should also be uint32 values, so no negative fill. We should discuss how to handle these datatypes more generally across all instruments...

File: imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml
Comment (greglucas) [inline imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml]: ns?

File: imap_processing/cli.py
Comment (greglucas) [inline imap_processing/cli.py]: Remove the comment since it is addressed now.
```suggestion
```

File: imap_processing/lo/l0/lo_science.py
Comment (greglucas) [inline imap_processing/lo/l0/lo_science.py]: You don't need to do this now, but something to think about since you're in the refactor mode.

I think you could rework this to be:
1. Get an array of all your ints `raw_ints = [int(bin_str[i : i + bits_per_index], 2) for i in range(section_start, section_start + section_length, bits_per_index)]`
2. Then call `decompress_int(raw_ints, decompress, DECOMPRESSION_TABLES)` by passing in the full array.

The reason I suggest that is currently you've got the function call in the hot-loop and there is a lot of extra verification/validation going on there. I think you could then switch over to numpy broadcasting/array lookup and do all of the decompression in one call. (I think SWAPI or SWE has an example of this with `np.take`, or just `lookup_table[raw_ints]` to do fancy indexing on your lookup table array.

File: imap_processing/lo/l1a/lo_l1a.py
Comment (greglucas) [inline imap_processing/lo/l1a/lo_l1a.py]: Rather than deleting them, what about only indexing the return back out of this function with the ones you can test?

```
good_apids = [LoAPID.ILO_SCI_CNT]  # Add to this as you go
return [datasets_by_apid[good_apid] for good_apid in good_apids])
```

File: imap_processing/lo/l1a/lo_l1a.py
Comment (greglucas) [inline imap_processing/lo/l1a/lo_l1a.py:142]: Would it be helpful to add a flag to remove the header from the packet_file_to_datasets return? I think @tech3371 suggested that as well, so maybe this is an indication it would be nice for multiple people.

File: imap_processing/lo/packet_definitions/lo_xtce.xml
Comment (greglucas) [inline imap_processing/lo/packet_definitions/lo_xtce.xml]: Are you manually updating these, or did you get a new spreadsheet? Just wondering if we're going to end up overwriting this again in the future possibly and need to make a note.
```suggestion
			<xtce:Parameter name="ILO_SCI_CNT.SCI_CNT" parameterTypeRef="ILO_SCI_CNT.SCI_CNT" shortDescription="Binary SCI CNT data" />
```

**Review by greglucas — COMMENTED** [review]
Leaving some initial feedback for now, likely won't be able to review again for a while, so don't wait for my final review/approval.

File: imap_processing/lo/l0/lo_science.py
Comment (tmplummer) [inline imap_processing/lo/l0/lo_science.py]: This global constant name should be all capitalized per PEP8: https://peps.python.org/pep-0008/#constants

**Review by tmplummer — COMMENTED** [review]
Looks good to me. I left one nit pick on naming. Ping me for review again once you address some of gregs comments.

File: imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml
Comment (laspsandoval) [reply imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml:45]: Agreed with Greg. Maybe at the next ENA tagup we can discuss common fillvals.

File: imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml
Comment (laspsandoval) [inline imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml:1]: I might be missing this, but where are the default_attrs used? For example, I used them like this for Ultra for x_front. Would it make sense to do something similar?

default_attrs: &default
  DEPEND_0: epoch
...

x_front:
  <<: *default
  CATDESC: x front position
...

File: imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml
Comment (laspsandoval) [inline imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml:43]: VALIDMIN, VALIDMAX, and FORMAT Values: Since many fields share the same VALIDMIN, VALIDMAX, and FORMAT values, you could consider defining these values in a common section and reference them, similar to how you handle default attributes.

**Review by laspsandoval — COMMENTED** [review]
A few comments/suggestions.

File: imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml
Comment (sdhoyt) [reply imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml:1]: default is used in the histogram fields:
```python
start_a:
  <<: *default
  VALIDMIN: 0
  CATDESC: Anode A Singles count
  DEPEND_1: azimuth_60
  DEPEND_2: esa_step
  FIELDNAM: Anode A Singles Count
  FORMAT: I12
  LABLAXIS: anode A singles counts
```

but I do need to clean the yaml up. I'll go through this and try to make the defaults used more efficiently.

File: imap_processing/lo/l1a/lo_l1a.py
Comment (sdhoyt) [reply imap_processing/lo/l1a/lo_l1a.py:142]: I agree, I think having that flag would be helpful.

File: imap_processing/lo/packet_definitions/lo_xtce.xml
Comment (sdhoyt) [reply imap_processing/lo/packet_definitions/lo_xtce.xml]: I manually updated it in the XTCE, but also updated the spreadsheet. I'll fix that missed DE instance

File: imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml
Comment (sdhoyt) [reply imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml:43]: Agreed. I have the default variable, but I noticed  I'm still repeating things when I don't need to. There are other changes I could make to be more efficient as well. Making this update will affect all the fields I currently have in L1A and might be a fairly large change to my yaml file. Would you be ok with me doing this is a separate PR so this one doesn't increase in size?

**Comment by tmplummer** [comment]
@sdhoyt, it is not clear to me what comments have been addressed or that you feel are resolved. Can you mark what you think is done as resolved and then ping me and I will review.

File: imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml
Comment (sdhoyt) [reply imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml:45]: I'm going to create a separate PR to address a bunch of CDF YAML issues: https://github.com/IMAP-Science-Operations-Center/imap_processing/issues/843

File: imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml
Comment (sdhoyt) [reply imap_processing/cdf/config/imap_lo_l1a_variable_attrs.yaml:43]: Going to fix the comments related to CDF definitions in a new PR for this issue: https://github.com/IMAP-Science-Operations-Center/imap_processing/issues/843

**Comment by sdhoyt** [comment]
> @sdhoyt, it is not clear to me what comments have been addressed or that you feel are resolved. Can you mark what you think is done as resolved and then ping me and I will review.

@subagonsouth All the comments are now marked as resolved. Most of the CDF YAML related comments are resolved by the creation of this issue: https://github.com/IMAP-Science-Operations-Center/imap_processing/issues/843. I'm going to make improvements to the YAML in a separate PR.

**Review by tmplummer — APPROVED** [review]
Sorry for the slow feedback.

**Pull request merged**
